mls-rs/src/tree_kem/math.rs

```diff
@@ -3,121 +3,132 @@
-    let mut m = 1 << (level(x) + 1);
+#[derive(Clone, PartialEq, Eq, core::fmt::Debug)]
+pub struct CopathNode<T: Clone + PartialEq + Eq + core::fmt::Debug> {
+    pub path: T,
+    pub copath: T,
+}
 
-    while m <= n {
-        d.push((x & !m) | (m - 1));
-        m <<= 1;
+impl<T: Clone + PartialEq + Eq + core::fmt::Debug> CopathNode<T> {
+    pub fn new(path: T, copath: T) -> CopathNode<T> {
+        CopathNode { path, copath }
     }
+}
 
-    Ok(d)
+#[derive(Clone, PartialEq, Eq, core::fmt::Debug)]
+pub struct ParentSibling<T: Clone + PartialEq + Eq + core::fmt::Debug> {
+    pub parent: T,
+    pub sibling: T,
 }
 
-pub fn copath(mut x: u32, n: u32) -> Result<Vec<u32>, MlsError> {
-    if x > 2 * n - 1 {
-        return Err(MlsError::InvalidTreeIndex);
+impl<T: Clone + PartialEq + Eq + core::fmt::Debug> ParentSibling<T> {
+    pub fn new(parent: T, sibling: T) -> ParentSibling<T> {
+        ParentSibling { parent, sibling }
     }
+}
 
-    let mut d = Vec::new();
+macro_rules! impl_tree_stdint {
+    ($t:ty) => {
+        impl TreeIndex for $t {
+            fn root(&self) -> $t {
+                *self - 1
+            }
 
-    while x != root(n) {
-        let p = parent(x);
+            /// Panicks if `x` is even.
+            fn left_unchecked(&self) -> Self {
+                *self ^ (0x01 << (level(*self) - 1))
+            }
 
-        d.push(if x < p {
-            right_unchecked(p)
-        } else {
-            left_unchecked(p)
-        });
+            /// Panicks if `x` is even.
+            fn right_unchecked(&self) -> Self {
+                *self ^ (0x03 << (level(*self) - 1))
+            }
 
-        x = p;
-    }
+            fn parent_sibling(&self, leaf_count: &Self) -> Option<ParentSibling<Self>> {
+                if self == &leaf_count.root() {
+                    return None;
+                }
 
-    Ok(d)
-}
+                let lvl = level(*self);
+                let p = (self & !(1 << (lvl + 1))) | (1 << lvl);
 
-pub fn path_copath(mut x: u32, n: u32) -> Result<Vec<(u32, u32)>, MlsError> {
-    if x > 2 * n - 1 {
-        return Err(MlsError::InvalidTreeIndex);
-    }
+                let s = if *self < p {
+                    p.right_unchecked()
+                } else {
+                    p.left_unchecked()
+                };
 
-    let mut d = Vec::new();
+                Some(ParentSibling::new(p, s))
+            }
 
-    while x != root(n) {
-        let p = parent(x);
+            fn is_leaf(&self) -> bool {
+                self & 1 == 0
+            }
 
-        let s = if x < p {
-            right_unchecked(p)
-        } else {
-            left_unchecked(p)
-        };
+            fn is_in_tree(&self, root: &Self) -> bool {
+                *self <= 2 * root
+            }
 
-        d.push((p, s));
-        x = p;
-    }
+            fn zero() -> Self {
+                0
+            }
+        }
 
-    Ok(d)
+        fn level(x: $t) -> u32 {
+            x.trailing_ones()
+        }
```

Comment:
[Nit] This will be problematic (duplicate name) if we ever want to invoke this macro twice in the same module. It's fine for now though.

Reply:
Yeah, that's why there's a `mod` around `impl_tree_stdint!(u64)`. Removed now